hnn_core/parallel_backends.py

```
@@ -0,0 +1,304 @@
+
+    def __enter__(self):
+        global BACKEND
+
+        self._old_backend = BACKEND
+        BACKEND = self
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        global BACKEND
+
+        BACKEND = self._old_backend
+
+    def _clone_and_simulate(self, net, trial_idx):
+        # avoid relative lookups after being forked by joblib
+        from hnn_core.neuron import NeuronNetwork, _simulate_single_trial
+
+        if trial_idx != 0:
+            net.params['prng_*'] = trial_idx
+
+        neuron_net = NeuronNetwork(net)
+        dpl = _simulate_single_trial(neuron_net)
+
+        spikedata = neuron_net.get_data_from_neuron()
+
+        return dpl, spikedata
+
+    def simulate(self, net):
+        """Simulate the HNN model
+
+        Parameters
+        ----------
+        net : Network object
+            The Network object specifying how cells are
+            connected.
+
+        Returns
+        -------
+        dpl: list of Dipole
+            The Dipole results from each simulation trial
+        """
+
+        n_trials = net.params['N_trials']
+        dpls = []
+
+        parallel, myfunc = self._parallel_func(self._clone_and_simulate)
+        sim_data = parallel(myfunc(net, idx) for idx in range(n_trials))
+
+        dpls = _gather_trial_data(sim_data, net, n_trials)
+        return dpls
+
+
+class MPIBackend(object):
+    """The MPIBackend class.
+
+    Parameters
+    ----------
+    n_procs : int | None
+        The number of processes processes MPI will use (spread over cores)
+    mpi_cmd : str
+        The name of the mpi launcher executable. Will use 'mpiexec'
+        (openmpi) by default.
+
+    Attributes
+    ----------
+
+    n_jobs : int
+        The number of jobs to start in parallel (NOT SUPPORTED)
+    n_procs : int
+        The number of processes MPI will use (spread over cores)
+    mpi_cmd_str : str
+        The string of the mpi command with number of procs and options
```

if the user provides `n_procs` and then it's also provided here, what will happen? Should we raise an error or MPI will take care of it?

---

I cleaned up the docstring to explain that one was user-specified, and the other is how many MPI will actually use. NEURON will print how many it is running on, so I think that is sufficient. If the user is seeing a discrepancy, they would be able to refer to the docstring. I don't think it needs to be explained in the output or with a warning.